### DISCUSSION TITLE
Test driven development on the library

### DISCUSSION THREAD

**Discussion by ingo-h** [pupnp/pupnp#282] [Ideas]
I have made a suggestion to use test driven development (TDD) on the library located in directory `upnp/gtest`. As seen by discussions on other comments it seems there is some detailed information needed to understand what I mean how to use Unit Tests. I will refer to [Wikipedia - Test-driven development](https://en.wikipedia.org/wiki/Test-driven_development). To have a common discussion base please also refer to this document.

TDD is part of **extreme programming** and more than running automatically tests on an existing code base. As stated at Point 1. "*The adding of a new feature begins by writing a test*". That's an essential paradigm of TDD. The test is made before the code. or in other words, in the beginning the test will always fail because there is no code under test. Now following the other Points 2. to 5. there may be a longer development process where tests will intentionally fail.

Unit tests are an integral and interactive part of programming. Requiring to pass the Unit tests on pull requests and on merging has the following disadvantages:

 1. A developer cannot provide intermediate steps of his code base for the time he need to complete his work.
 2. He cannot provide a test for a bug so other developers cannot run the test and inspect its conditions, for example https://github.com/pupnp/pupnp/issues/247 and https://github.com/pupnp/pupnp/issues/272.
 3. He cannot show that a test is running successful because another test fails. Here I could show that `UpnpGetIfInfo(p, "if0v4")` (https://github.com/pupnp/pupnp/pull/271#issuecomment-808930231) is successful running but the sanitity-check fails because of https://github.com/pupnp/pupnp/issues/247.
4. The sanity-check is not the right environment to perform development as noted by Point 1. to 5. at [Test-driven development cycle](https://en.wikipedia.org/wiki/Test-driven_development#Test-driven_development_cycle).
5. It is unclear how the gtests are executed on the sanity-checks. You need either **googletest** and **googlemock** and the testsuites are compiled, linked and executed. Then it is the question, why only the `test_upnpapi.cpp` testsuite is executed and not the `test_template.cpp`. Or there is an old precompiled test `test_upnpapi.a` running, that I forgot to delete. Because of the static nature of the tests that does not ensure to match the current state of the library.

On the other hand it doesn't hurt to fiddle with the unit tests and use them to show other developers with simple tests what is meant instead of describing things in comments that cannot be executed and verified. Due to the nature of Unit tests they are completely independent from the production code and will normally not distributed with it. The `upnp/gtest` directory can be seen as an independent subdirectory not involving the production code.

At the beginning I had provided an example installation. That is what I need to have a proper Unit-test environment with effective collaboration:

 1. **googletest** and **googlemock** online downloaded, compiled and installed. This is already done and is working very well.
 2. A gtest subdirectory that isn't involved into the normal install, configure and build process. As shown above it is not part of the production process. It is only part of the development and should be seen like a `src/` directory. There you will not find any cmake management files. The same should be with the `gtest/` directory.
 3. Never running the gtest testsuites as sanity-checks on pull-requests and merges.
 3. It would be nice to have the gtests linked against the shared upnp library, so you can create an executable test one time and only modify the library to fit to the test step by step. But this needs **googletest** to be relocatable compiled. It isn't direct supported and with a first glance at the sources of **googletest** I saw that it isn't done for me in day.

**Comment by mrjimenez** [comment]
Hi Ingo,

I think we both agree that TDD is not Unit Testing. Sanity tests are unit testing. They make sure that we don't have a regression in the code. That is not to say that I don't test my code, but e.g., I have no Mac and no Windows to test, so the Github actions come in handy. Also I understand that the tests used in TDD can later be used as sanity tests and regression control.

I believe I am being repetitive, but let me say that again: 
- I like the testing infra-structure that you are working hard to put in.
- This is one of the reasons why libupnp now admits C++. A rather strong reason, btw.
- I did install google test and google mock by hand in my system so that I can use it.
- I am willing to learn how to create tests using it.

That said, you have a pull request that I am more than willing to commit, but has not been committed because it is currently being used as unit test. Just disable it and we're done, your code is in.

After that commit, what I would really like is that we have two "environments". One for TDD, the other for Unit tests, both based on the test infra-structure that you are doing.

Regards,
Marcelo.

**Comment by Vollstrecker** [comment]
Disadvantage 1: Why? As you see you can mark a test as will fail. In this case the failure of the test is treated as success. So everyone (at least if he follows the process) knows, as long as all succeed the feature is not done. When the development is done the test, you just remove the WILL_FAIL flag and everything is OK. The test remains active and assures that the finished feature won't be broken again.

Disadvantage 2: Why? For just showing, a good description is in complex scenarios the best way. In #247 everyone understood after your first post what's the problem. When I dig trough your code, I will find out what you're doing, but this will take more time than just reading this view lines. 

Disadvantage 3:  Again Why? You'll never be able to proof something doesn't work, if you don't give it the chance to.If you want to demonstrate that destroying a mutex crashes in some specific case, you are responsible to create one first.

Disadvantage 4: Everything is absolutely doable. as described for Point 1 it's no problem to write a test that fails. A test that is expected to fail doesn't have to count as fail, but regarding how the result is treated, there is nothing stated there.

Disadvantage 5: What are you missing? First: You have to represent the current state of the lib, as you link against it. And yes this is mandatory, otherwise you test maybe long abandoned features. Second: Static is only gmock, the test are created 2 times, for static and dynamic linking. You see that you have testname and testname-static. For test_template: you include it in the apitest, so it should be there. To get the list of tests the code is parsed, maybe the included code it not. If test_template is useable on it's own, you an just add it to the tests and it will be executed. If it's just helper-clases that always need the apitests, then defining tests in there seems to me not the right place and you should think about moving them either into apitests or into a new file.

Remaining points:
1: This is done, but downloading should always be a fallback. I don't see any advantage over just installing the packages your distro provides.

2: The files in src/ get included. I could add an option to not include them, but to be honest, if the test are done correctly, I don't see a reason to exclude them. OTOH you added the downloading of googletest to the main file, and it's a real bad idea to unconditionally download and build stuff that isn't used by the build.

3: If you don't do that, you maybe run these tests until the problem is fixed. Why should anyone run them again? Including them is the only way to ensure that the fix isn't broken later.

4: The tests itself are also linked against the shared upnp. Admitted, it's a mix of shared upnp and static googletest, but this shouldn't really be a problem. Building it relocatable isn't needed because they are not relacated, the are just used on there place and never leave it. The bigger problem is that googletest was not designed to provide both versions at the same time. I could build it twice, but as they both use the same targetname this can't work. But I can add a switch, so you can decide which one you want to use.
